File: src/hashcode.py
import sys
import os


def parsing(filename: str):
    """ Parse the input file.

    Args:
        filename: file to parse

    Returns:
        meta data and tab
    """
    with open(filename) as f:
        rows, cols, min_ing, max_area = [ int(v) for v in f.readline().split(' ')]
        tab = [ [ cell for cell in line.strip() ] for line in f.readlines() ]
    meta = {'rows': rows, 'cols': cols,
            'min_ing': min_ing, 'max_area': max_area}
    return meta, tab


def get_forms(min_area: int, max_area: int):
    """ Make all the form of slice.

    Args:
        min_area: the minimal area required
        max_area: the maximal area possible

    Returns:
        list of forms
    """
    forms = [ (x, y) for y in range(1, max_area + 1)
                        for x in range(1, max_area + 1)
                        if min_area <= x * y <= max_area ]
    return forms


def is_valid(pos, form, min_ing, tab, size):
    pos_x, pos_y = pos
    tab_x, tab_y = size
    form_x, form_y = form
    m, t = (0, 0)
    if pos_x + form_x > tab_x or pos_y + form_y > tab_y :
        return 0
    for x in range(form_x):
        for y in range(form_y):
            ingredient = tab[pos_y + y][pos_x + x]
            if ingredient == 'M':
                m += 1
            elif ingredient == 'T':
                t += 1
            else:
                raise ValueError
            if m >= min_ing and t >= min_ing:
                return 1
    return 0


def counter(tab_ref, tab_counter, forms, meta):
    tab_x, tab_y = meta['cols'], meta['rows']
    for x, y in [(x, y) for y in range(tab_y) for x in range(tab_x)]:
        for form in forms:
            form_x, form_y = form
            if not is_valid((x,y), form, meta['min_ing'], tab_ref,
                    (meta['cols'], meta['rows'])):
                continue
            for pos_x, pos_y in [ (pos_x, pos_y) for pos_x in range(form_x) for
                    pos_y in range(form_y)]:
                tab_counter[pos_y + y][pos_x + x] += 1
    for line in tab_counter:
        print(line)
    return tab_counter


if __name__ == '__main__':
    if len(sys.argv) != 2:
        print('usage: python3 {:s} file'.format(sys.argv[0]))
        exit(os.EX_DATAERR)
    meta, tab = parsing(sys.argv[1])
    forms = get_forms(2 * meta['min_ing'], meta['max_area'])
    print('#   META   #')
    print(meta)
    print('')
    print('#   FROMS   #')
    for form in forms:
        print(form)
    print('')
    print('#   TAB   #')
    for lines in tab:
        print(lines)
    print('')
    print('#   COUNTER   #')
    counter(tab, [ [ 0 for x in range(meta['cols']) ] for y in
        range(meta['rows']) ], forms, meta)

File: src/hashcode.js
const fs = require('fs')
const _ = require('lodash')

const parsing = file => {
  const raw = fs.readFileSync(file, 'utf8').split('\n').slice(0, -1)
  const [rows, cols, minIng, maxArea] = raw[0].split(' ').map(n => parseInt(n, 10))
  const data = raw.slice(1).map(row => row.split(''))

  return {
    meta: { rows, cols, minIng, maxArea },
    data,
  }
}

const getForms = (min, max) => {
  const range = _.range(1, max + 1)

  return _.filter(_.flatMap(range, y =>
    _.map(range, x => {
      if (min <= (x * y) && (x * y) <= max)
        return [x, y]
    })
  ), undefined)
}

const isValid = ({ posX, posY }, { formX, formY }, minIng, data, { tabX, tabY }) => {
  if (posX + formX > tabX || posY + formY > tabY)
    return 0

  const [m, t] = [0, 0]

  for (const x of _.range(formX)) {
    for (const y of _.range(formY)) {
      const item = data[posY + y][posX + x]

      if (!['M', 'T'].includes(item))
        throw new Error('Value error')

      m += _.isEqual(item, 'M') ? 1 : 0
      t += _.isEqual(item, 'T') ? 1 : 0

      if (m >= minIng && t >= minIng)
        return 1
    }
  }
}

const makeCoords = (maxX, maxY) => {
  const coords = []

  for (const y of _.range(maxY)) {
    for (const x of _.range(maxX)) {
      coords.push([x, y])
    }
  }

  return coords
}

(() => {
  try {
    const script = __filename.split('/').slice(-1)[0]
    const file = process.argv[2]

    if (!file) throw new Error(`Usage: node ${script} file`)

    const { meta, data } = parsing(file)

    const forms = getForms(2 * meta.minIng, meta.maxArea)

  } catch (err) {
    console.error(err)
    process.exit(1)
  }
})()
